File: sdks/rust/qnsi/src/audit.rs
//! QNSI Audit — immutable, hash-chained event log.

use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::Arc;

use crate::activation::Activation;
use crate::errors::Error;
use crate::service_client::ServiceClient;

const PATH_PREFIX: &str = "/proxy/audit/v1";

#[derive(Clone)]
pub struct Client {
    sc: ServiceClient,
}

impl Client {
    pub fn new(activation: Arc<Activation>, http: reqwest::Client) -> Self {
        Self { sc: ServiceClient::new(activation, http, PATH_PREFIX) }
    }

    /// Append a single event. `POST /audit/v1/events` (the event object directly).
    pub async fn log_event(
        &self,
        req: LogEventRequest,
        idempotency_key: Option<&str>,
    ) -> Result<Value, Error> {
        self.sc
            .request(Method::POST, "/events", Some(&req), None, idempotency_key)
            .await
    }

    /// Append a batch of events. `POST /audit/v1/events/batch` with `{ events }`.
    pub async fn ingest_events(
        &self,
        events: Vec<LogEventRequest>,
        idempotency_key: Option<&str>,
    ) -> Result<Value, Error> {
        #[derive(Serialize)]
        struct Body<'a> {
            events: &'a [LogEventRequest],
        }
        self.sc
            .request(
                Method::POST,
                "/events/batch",
                Some(&Body { events: &events }),
                None,
                idempotency_key,
            )
            .await
    }

    pub async fn list_events(&self, query: &[(&str, String)]) -> Result<Value, Error> {
        self.sc
            .request::<()>(Method::GET, "/events", None, Some(query), None)
            .await
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct LogEventRequest {
    #[serde(rename = "eventType")]
    pub event_type: String,
    pub payload: serde_json::Map<String, Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
}

File: sdks/go/qnsi/audit/audit.go
// Package audit is the QNSP Audit client - immutable, hash-chained event log.
package audit

import (
	"context"
	"net/http"
	"time"

	"github.com/heossihq/qnsi-public/sdks/go/qnsi/internal/qnsicore"
)

const pathPrefix = "/proxy/audit/v1"

type Client struct {
	base *qnsicore.ServiceClient
}

func New(act *qnsicore.Activator, httpClient *http.Client, timeout time.Duration) *Client {
	return &Client{base: qnsicore.NewServiceClient(act, httpClient, pathPrefix, timeout)}
}

type LogEventRequest struct {
	EventType string         `json:"eventType"`
	Payload   map[string]any `json:"payload"`
	Tags      []string       `json:"tags,omitempty"`
}

func (c *Client) LogEvent(ctx context.Context, req LogEventRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/events", req, nil, idempotencyKey)
}

func (c *Client) IngestEvents(ctx context.Context, events []LogEventRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/events/batch", map[string]any{"events": events}, nil, idempotencyKey)
}

func (c *Client) ListEvents(ctx context.Context, query map[string]string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/events", nil, query, "")
}
